From f83bdf22acc8d2e5e2d88bef888d6ba3a92397a4 Mon Sep 17 00:00:00 2001
From: Alex Wilson <alex.wilson@joyent.com>
Date: Mon, 30 Apr 2018 13:20:07 -0700
Subject: [PATCH] MANTA-3672 want support for cross-account RBAC role
 membership Reviewed by: Cody Peter Mello <cody.mello@joyent.com> Approved by:
 Cody Peter Mello <cody.mello@joyent.com>

---
 lib/server/server.js          | 10 +++++++---
 test/data/test-nodeletes.json |  1 +
 test/data/test-nodeletes.ldif |  8 ++++++++
 test/server.test.js           | 28 ++++++++++++++++++++++++++++
 4 files changed, 44 insertions(+), 3 deletions(-)

diff --git a/lib/server/server.js b/lib/server/server.js
index e1c8031..4cc7308 100644
--- a/lib/server/server.js
+++ b/lib/server/server.js
@@ -163,12 +163,12 @@ function Server(opts) {
     server.get({
         name: 'getAccountByUuid',
         path: '/accounts/:accountid'
-    }, [getAccount, sendAuth]);
+    }, [getAccount, getRoles, sendAuth]);
 
     server.get({
         name: 'getAccount',
         path: '/accounts'
-    }, [getAccountUuid, getAccount, sendAuth]);
+    }, [getAccountUuid, getAccount, getRoles, sendAuth]);
 
     server.get({
         name: 'getUserByUuid',
@@ -383,7 +383,11 @@ function getUser(req, res, next) {
 
 
 function getRoles(req, res, next) {
-    var roles = req.auth.user.roles || [];
+    var roles;
+    if (req.auth.user)
+        roles = req.auth.user.roles || [];
+    else
+        roles = req.auth.account.roles || [];
     req.log.debug({roles: roles}, 'getRoles handler: entered');
     lib.getRoles({
         roles: roles,
diff --git a/test/data/test-nodeletes.json b/test/data/test-nodeletes.json
index 07320f4..25e0873 100644
--- a/test/data/test-nodeletes.json
+++ b/test/data/test-nodeletes.json
@@ -36,3 +36,4 @@
 {"dn":"changenumber=38, cn=changelog","controls":[],"targetdn":"group-uuid=f7e61afc-9f12-11e3-8353-4b17a217ad14, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc","changetype":"modify","objectclass":"changeLogEntry","changetime":"2014-02-27T19:53:38.181Z","changes":"[{\"operation\":\"delete\",\"modification\":{\"type\":\"memberpolicy\",\"vals\":[\"policy-uuid=a188ee70-9fdd-11e3-925f-ab3b4c32f7b8, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"]}}]","entry":"{\"account\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"name\":[\"borrower\"],\"objectclass\":[\"sdcaccountrole\"],\"uuid\":[\"f7e61afc-9f12-11e3-8353-4b17a217ad14\"],\"_owner\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"_parent\":[\"uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"],\"memberpolicy\":[\"policy-uuid=b4301b32-66b4-11e3-ac31-6b349ce5dc45, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"]}","changenumber":"38"}
 {"dn":"changenumber=39, cn=changelog","controls":[],"targetdn":"policy-uuid=b4301b32-66b4-11e3-ac31-6b349ce5dc45, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc","changetype":"modify","objectclass":"changeLogEntry","changetime":"2014-02-27T19:53:38.266Z","changes":"[{\"operation\":\"replace\",\"modification\":{\"type\":\"rule\",\"vals\":[\"Can read x and y when ip = 10.0.0.0/32\"]}}]","entry":"{\"account\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"name\":[\"foo\"],\"objectclass\":[\"sdcaccountpolicy\"],\"rule\":[\"Can read x and y when ip = 10.0.0.0/32\"],\"uuid\":[\"b4301b32-66b4-11e3-ac31-6b349ce5dc45\"],\"_owner\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"_parent\":[\"uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"],\"memberrole\":[\"group-uuid=f7e61afc-9f12-11e3-8353-4b17a217ad14, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"]}","changenumber":"39"}
 {"dn":"changenumber=40, cn=changelog","controls":[],"targetdn":"policy-uuid=b4301b32-66b4-11e3-ac31-6b349ce5dc45, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc","changetype":"modify","objectclass":"changeLogEntry","changetime":"2014-02-27T19:53:38.280Z","changes":"[{\"operation\":\"add\",\"modification\":{\"type\":\"rule\",\"vals\":[\"Can read a and b when ip = 10.0.0.0/32\"]}}]","entry":"{\"account\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"name\":[\"foo\"],\"objectclass\":[\"sdcaccountpolicy\"],\"rule\":[\"Can read a and b when ip = 10.0.0.0/32\",\"Can read x and y when ip = 10.0.0.0/32\"],\"uuid\":[\"b4301b32-66b4-11e3-ac31-6b349ce5dc45\"],\"_owner\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"_parent\":[\"uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"],\"memberrole\":[\"group-uuid=f7e61afc-9f12-11e3-8353-4b17a217ad14, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"]}","changenumber":"40"}
+{"dn":"changenumber=41, cn=changelog","controls":[],"targetdn":"group-uuid=fd4d1489-a2c4-4303-8b32-0396ca297447, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc","changetype":"add","objectclass":"changeLogEntry","changetime":"2014-02-27T19:53:36.888Z","changes":"{\"account\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"name\":[\"crossrole\"],\"objectclass\":[\"sdcaccountrole\"],\"uniquemember\":[\"uuid=2a05359a-9e64-11e3-816d-e7f87365cf40, ou=users, o=smartdc\"],\"uuid\":[\"fd4d1489-a2c4-4303-8b32-0396ca297447\"],\"_owner\":[\"bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f\"],\"_parent\":[\"uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc\"]}","changenumber":"41"}
diff --git a/test/data/test-nodeletes.ldif b/test/data/test-nodeletes.ldif
index c8c77b1..203eebe 100644
--- a/test/data/test-nodeletes.ldif
+++ b/test/data/test-nodeletes.ldif
@@ -183,6 +183,14 @@ account: 2a05359a-9e64-11e3-816d-e7f87365cf40
 uniquemember: uuid=a40a2764-9e65-11e3-85e0-33cd72381f16, uuid=2a05359a-9e64-11e3-816d-e7f87365cf40, ou=users, o=smartdc
 objectclass: sdcaccountrole
 
+# account: banks
+dn: group-uuid=fd4d1489-a2c4-4303-8b32-0396ca297447, uuid=bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f, ou=users, o=smartdc
+changetype: add
+name: crossrole
+uuid: fd4d1489-a2c4-4303-8b32-0396ca297447
+account: bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f
+uniquemember: uuid=2a05359a-9e64-11e3-816d-e7f87365cf40, ou=users, o=smartdc
+objectclass: sdcaccountrole
 
 ## --- sdcaccountpolicy add
 
diff --git a/test/server.test.js b/test/server.test.js
index 2f260a9..c5b1a6e 100644
--- a/test/server.test.js
+++ b/test/server.test.js
@@ -192,6 +192,34 @@ test('get user by login', function (t) {
     });
 });
 
+test('cross-account roles', function (t) {
+    var uuid = '2a05359a-9e64-11e3-816d-e7f87365cf40';
+    this.client.get('/accounts/' + uuid, function (err, req, res, obj) {
+        t.ok(obj.account);
+        t.ok(obj.roles);
+        t.deepEqual(obj.roles, {
+            'fd4d1489-a2c4-4303-8b32-0396ca297447': {
+                type: 'role',
+                uuid: 'fd4d1489-a2c4-4303-8b32-0396ca297447',
+                name: 'crossrole',
+                account: 'bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f',
+                rules: []
+            }
+        });
+        t.end();
+    });
+});
+
+test('cross-account roles (none)', function (t) {
+    var uuid = 'bde5a308-9e5a-11e3-bbf2-1b6f3d02ff6f';
+    this.client.get('/accounts/' + uuid, function (err, req, res, obj) {
+        t.ok(obj.account);
+        t.ok(obj.roles);
+        t.deepEqual(obj.roles, {});
+        t.end();
+    });
+});
+
 test('translate account', function (t) {
     this.client.get('/uuids?account=banks', function (err, req, res, obj) {
         t.ok(obj.account);
-- 
2.21.0

